File: Airship/AirshipCore/Source/Internal/UADeviceRegistrationEvent.m
/* Copyright Airship and Contributors */

#import "UADeviceRegistrationEvent+Internal.h"
#import "UAEvent+Internal.h"
#import "UAPush.h"
#import "UAChannel.h"
#import "UAirship.h"

#if __has_include("AirshipCore/AirshipCore-Swift.h")
#import <AirshipCore/AirshipCore-Swift.h>
#elif __has_include("Airship/Airship-Swift.h")
#import <Airship/Airship-Swift.h>
#endif

@implementation UADeviceRegistrationEvent

+ (instancetype)event {
    UADeviceRegistrationEvent *event = [[self alloc] init];

    NSMutableDictionary *data = [NSMutableDictionary dictionary];

    if ([[UAirship shared].privacyManager isEnabled:UAFeaturesPush]) {
        [data setValue:[UAirship push].deviceToken forKey:@"device_token"];
    }

    [data setValue:[UAirship channel].identifier forKey:@"channel_id"];

    event.eventData = data;
    return event;
}

- (NSString *)eventType {
    return @"device_registration";
}

@end

File: Airship/AirshipCore/Source/UADeviceRegistrationEvent.swift
/* Copyright Airship and Contributors */

/**
 * @note For Interrnal use only :nodoc:
 */
@objc
public class UADeviceRegistrationEvent : NSObject, UAEvent {
    
    @objc
    public init(channel: ChannelProtocol?,
                push: PushProtocol?,
                privacyManager: UAPrivacyManager?) {
         
        var data: [AnyHashable : Any] = [:]
        if privacyManager?.isEnabled(.push) == true {
            data["device_token"] = push?.deviceToken
        }

        data["channel_id"] = channel?.identifier

        self._data = data
    }
    
    @objc
    public override convenience init() {
        self.init(channel: UAirship.channel(), push: UAirship.push(), privacyManager: UAirship.shared().privacyManager)
    }
    
    @objc
    public var priority: UAEventPriority {
        get {
            return .normal
        }
    }
    
    @objc
    public var eventType : String {
        get {
            return "device_registration"
        }
    }

    private let _data : [AnyHashable : Any]

    @objc
    public var data: [AnyHashable : Any] {
        get {
            return self._data
        }
    }
}
